nemo/core/classes/exportable.py

```diff
@@ -0,0 +1,143 @@
+            input_names = list(self.input_types.keys())
+            output_names = list(self.output_types.keys())
+            dynamic_axes = defaultdict(list)
+
+            # extract dynamic axes and remove unnecessary inputs/outputs
+            # for input_ports
+            for _name, ntype in self.input_types.items():
+                if _name in self.disabled_deployment_input_names:
+                    input_names.remove(_name)
+                    continue
+                self.__extract_dynamic_axes(_name, ntype, dynamic_axes)
+            # for output_ports
+            for _name, ntype in self.output_types.items():
+                if _name in self.disabled_deployment_output_names:
+                    output_names.remove(_name)
+                    continue
+                self.__extract_dynamic_axes(_name, ntype, dynamic_axes)
+
+            if len(dynamic_axes) == 0:
+                dynamic_axes = None
+
+            # Set module to eval mode
+            self.eval()
+
+            # Attempt export
+            if format == ExportFormat.ONNX:
+                if _in_example is None:
+                    raise ValueError(f'Example input is None, but ONNX tracing was attempted')
+                if _out_example is None:
+                    if isinstance(_in_example, tuple):
+                        _out_example = self.forward(*_in_example)
+                    else:
+                        _out_example = self.forward(_in_example)
+                with torch.jit.optimized_execution(True):
+                    jitted_model = torch.jit.trace(self, _in_example)
+
+                torch.onnx.export(
+                    jitted_model,
+                    _in_example,
+                    output,
+                    input_names=input_names,
+                    output_names=output_names,
+                    verbose=False,
+                    export_params=True,
+                    do_constant_folding=True,
+                    dynamic_axes=dynamic_axes,
+                    opset_version=onnx_opset_version,
+                    example_outputs=_out_example,
+                )
+            else:
+                raise ValueError(f'Encountered unknown export format {format}.')
+        finally:
+            typecheck.set_typecheck_enabled(enabled=True)
+
+    @property
+    def disabled_deployment_input_names(self):
+        """Implement this method to return a set of input names disabled for export"""
+        return set()
+
+    @property
+    def disabled_deployment_output_names(self):
+        """Implement this method to return a set of output names disabled for export"""
+        return set()
+
+    @staticmethod
+    def __extract_dynamic_axes(port_name, ntype: NeuralType, dynamic_axes: defaultdict):
+        if ntype.axes:
+            for ind, axis in enumerate(ntype.axes):
+                if axis.kind == AxisKind.Batch or axis.kind == AxisKind.Time:
```

Comment:
Maybe add a todo here to add support for Width and Height for image models?

Reply:
actually, Width and Height aren't dynamic. The "dynamic" means it can vary from batch to batch. But I guess that could be true for width and height of some detection models ...

Reply:
We absolutely should. For ONNX purposes, 'dynamic' should be any dimension that may be different from run to run.

Reply:
reworked _extract_dynamic_axes a little bit. Also added comments to that people can override it if they need other dynamic axes.